tests/mpmc_channel.rs

```diff
@@ -754,6 +754,78 @@ macro_rules! gen_mpmc_tests {
                 assert_eq!(1, elem2.strong_count());
                 assert_eq!(1, elem3.strong_count());
             }
+
+            #[test]
+            fn buffered_uncompleted_recv_does_not_deadlock() {
+                let channel = ChannelType::new();
```

Comment:
Looking at the base test setup this creates a channel with a capacity of 3. So all those sends and receives can be served by the buffer. I would recommend adding 1-2 further concurrent transactions to first take up **all** permits. 

Sequence could be along:
5x receive_fut = channel.receive();
5x channel.send(); 3 of those should finish immediately. The others should be pending.

Now do 5x receive from **other** futures. All those should actually succeed. 3 can grab buffered data, 2 can grab hand-over data.

Do one other receive from yet another receiver. That will register a receiver a waiter again.
Do another send. The last waiter now should have been woken up.
Finish the last receive.

Then just doing 5xsend and 5x receive to clean up the inital waiters.

Reply:
Thanks for the update! So it doesn't run into the `debug_assert` - even though it directly copies from the sender? I'm surprised.